Compressed

jannikluhn opened the pull request:
### What was wrong?

Missing support for compressed public keys (#55).

### How was it fixed?

- Add `to/from_compressed_bytes` methods to `PublicKey` class.
- Add implementations for the coincurve and native backend.
- Add some tests.

I'm not an expert in cryptography, so please review carefully, especially the native backend code.

#### Cute Animal Picture

![Cute animal picture](https://user-images.githubusercontent.com/29854669/57235207-aa4eca00-7022-11e9-902f-2da640e6f689.jpg)

pipermerriam commented on eth_keys/backends/base.py:
```diff
@@ -27,3 +27,11 @@ def ecdsa_recover(self,
+
+    def decompress_public_key_bytes(self,
+                                    compressed_public_key_bytes: bytes) -> bytes:
+        raise NotImplementedError()
```
Note to self that we should convert these to be `ABC,abstractmethod` based sometime.

pipermerriam commented on eth_keys/validation.py:
```diff
@@ -56,10 +56,20 @@ def validate_message_hash(value: Any) -> None:
+    validate_bytes(value)
+    if len(value) != 33:
+        raise ValidationError(
+            "Unexpected compressed public key format.  Must be length 33 byte string"
```
Nice to include the actual value length here as well so it's clear what the discrepancy is.

jannikluhn commented:
Added the suggested hypothesis test, improved the validation error message, and updated the readme.

Also, when looking at the readme I realized that the key compression API is duplicated: Once in PublicKey (where one would naturally look for it) and once in `KeyAPI` (where it doesn't make too much sense IMO). I therefore removed it from `KeyAPI` and udpated the tests accordingly.

carver commented on eth_keys/backends/native/ecdsa.py:
```diff
@@ -60,6 +60,35 @@ def private_key_to_public_key(private_key_bytes: bytes) -> bytes:
+        prefix = b"\x02"
+    else:
+        prefix = b"\x03"
+    return prefix + pad32(int_to_big_endian(x))
```
lol, padding the compressed key is a strange spec. I was wondering why the compressed key is always the same length...

carver commented on tests/backends/test_native_backend_against_coincurve.py:
```diff
@@ -157,3 +147,62 @@ def test_coincurve_to_native_invalid_signatures(message_hash,
+@given(
+    private_key_bytes=private_key_st,
+    direction=st.one_of(
+        st.just('coincurve-to-native'),
```
What's the point of reversing direction on this? Looks like copy-pasta, to me.

jannikluhn replied:
Whoops, you're right, there's no point. Fixed.

carver commented:
Anything else to add today @pipermerriam ? If not, I'll merge by EOD.

pipermerriam commented on eth_keys/backends/coincurve.py:
```diff
@@ -67,3 +71,17 @@ def private_key_to_public_key(self, private_key: PrivateKey) -> PublicKey:
+                                  uncompressed_public_key_bytes: bytes) -> bytes:
+        point = (
+            big_endian_to_int(uncompressed_public_key_bytes[:32]),
+            big_endian_to_int(uncompressed_public_key_bytes[32:]),
```
This allows for an arbitrarily long second value were the provided `uncompressed_public_key_bytes` to be longer than `64` bytes.  Am I missing something?  Seems like this would cause weird/hard-to-debug downstream errors in some cases.

jannikluhn replied:
This is not part of the public API, all inputs should pass through `datatypes.PublicKey`. I still added validation now, just to be safe.

pipermerriam commented on eth_keys/backends/native/ecdsa.py:
```diff
@@ -60,6 +60,35 @@ def private_key_to_public_key(private_key_bytes: bytes) -> bytes:
+    if prefix not in (2, 3):
+        raise ValueError("Invalid compressed public key")
+
+    x = big_endian_to_int(compressed_public_key_bytes[1:])
```
I've been on a big `memoryview` kick recently so feel free to ignore me.  This would be slightly faster using a `memoryview` to prevent the copy.  This is only safe if the thing you're passing it to doesn't try to mutate it which in this case I believe that condition is satisfied.

jannikluhn replied:
I did a quick benchmark and didn't see any improvement. I think the reason is that it would save one copy when slicing, but it adds one when creating the `memoryview`. But I will keep it in mind in the future (could be very useful for pyrlp/ssz).

jannikluhn commented:
I don't have write access here, can someone merge please?

carver commented:
Would you like a release after #58 is merged?

jannikluhn commented:
Thanks!

> Would you like a release after #58 is merged?

Yep, that would be nice.
